### DISCUSSION TITLE
Importing pulsar-broker:2.9.3 or 2.10.1 maven dependency

### DISCUSSION THREAD

**Discussion by goncalocribeiro** [apache/pulsar#17988] [Q&A] [answered]
Hi everyone,

I'm struggling with a problem in my project and I hope someone here can help.

I'm implementing a custom authentication and authorization logic for my Pulsar implementation and currently I'm using the version 2.9.1.

Now, I want to upgrade to 2.9.3 but I found I have to update my implementation of AuthorizationProvider because the argument of the method waitForAny() changed from List to Collection.

The problem is that when importing the maven dependency "pulsar-broker" version 2.9.3, it just can't import the dependencies and stall infinitely (seems like circular dependency or something like that) , and the same happens with 2.10.1.
I have no problems with 2.9.1, 2.9.2 or 2.10.0.

Anyone with the same problem or any idea how to fix this?

Thanks

**Comment by HenriqueFSilva** [comment]
I have the same issue, would really like some feedback on this :)

**Comment by michaeljmarshall** [comment, marked as answer]
The break in binary compatibility was introduced here https://github.com/apache/pulsar/pull/15329, and fixed here, https://github.com/apache/pulsar/pull/16794. It should be fixed the 2.9.4 release, but that is not available yet. In order to use 2.9.3, you'll need to recompile the project, like you're trying to do.

As for being unable to download the dependency, that sounds like a separate issue. Do you have any logs? When you build the dependency tree, do you see a circular dependency? After quickly searching, I don't see any related issues.

**Reply by goncalocribeiro** [reply]
Hi Michael,

Actually I found out that I only need the pulsar-broker-common dependency, and by excluding the pulsar-zookeeper-utils dependency it works:

      <dependency>
          <groupId>org.apache.pulsar</groupId>
          <artifactId>pulsar-broker-common</artifactId>
          <version>2.9.3</version>
          <exclusions>
              <exclusion>
                  <groupId>org.apache.pulsar</groupId>
                  <artifactId>pulsar-zookeeper-utils</artifactId>
              </exclusion>
          </exclusions>
      </dependency>

I've tried before, to execute mvn dependency:tree, but the process just get stuck... so I couldn't get through that.

**Reply by michaeljmarshall** [reply]
Glad you found something that worked. When you enable debug logs, do you get any helpful messages to understand where it is stuck?

**Reply by goncalocribeiro** [reply]
Unfortunately no. I didn't find anything useful
